File: mainActivity/src/main/java/net/airvantage/utils/AuthenticationUrlParser.java
package net.airvantage.utils;

import java.util.Date;

import android.net.Uri;
import android.util.Log;

import com.sierrawireless.avphone.auth.Authentication;

public class AuthenticationUrlParser {

    public Authentication parseUrl(String url, Date parsingDate) {

        Authentication auth = null;

        if (url.startsWith("oauth")) {
            Log.d(AuthenticationUrlParser.class.getName(), "Callback URL: " + url);
            // Example
            // oauth://airvantage#access_token=430ad00a-2737-4673-bfa4-48c6710d748f&token_type=bearer&expires_in=86399&scope=
            Uri uri = Uri.parse(url);

            if (uri.getHost().equals("airvantage")) {
                auth = new Authentication();

                String fragment = uri.getFragment();
                if (fragment != null) {

                    String[] params = fragment.split("&");
                    for (String param : params) {
                        String[] kv = param.split("=");
                        if (kv != null && kv.length == 2) {
                            String key = kv[0];
                            String value = kv[1];

                            if ("access_token".equals(key)) {
                                auth.setAccessToken(value);
                            } else if ("expires_in".equals(key)) {
                                int expiresInSeconds = Integer.parseInt(value);
                                Date expirationDate = new Date(parsingDate.getTime() + expiresInSeconds * 1000);
                                auth.setExpirationDate(expirationDate);
                            }
                        }
                    }
                }
            }

        }

        return auth;
    }

}

File: mainActivity/src/main/java/net/airvantage/utils/AuthenticationUrlParser.kt
package net.airvantage.utils

import java.util.Date

import android.net.Uri
import android.util.Log

import com.sierrawireless.avphone.auth.Authentication

class AuthenticationUrlParser {

    private val TAG = this::class.java.name

    fun parseUrl(url: String, parsingDate: Date): Authentication? {

        var auth: Authentication? = null

        if (url.startsWith("oauth")) {
            Log.i(TAG, "Callback URL: " + url)
            // Example
            // oauth://airvantage#access_token=430ad00a-2737-4673-bfa4-48c6710d748f&token_type=bearer&expires_in=86399&scope=
            val uri = Uri.parse(url)

            if (uri.host == "airvantage") {
                auth = Authentication()

                val fragment = uri.fragment
                if (fragment != null) {

                    val params = fragment.split("&".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                    for (param in params) {
                        val kv = param.split("=".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                        if (kv.size == 2) {
                            val key = kv[0]
                            val value = kv[1]

                            if ("access_token" == key) {
                                auth.accessToken = value
                            } else if ("expires_in" == key) {
                                val expiresInSeconds = Integer.parseInt(value)
                                val expirationDate = Date(parsingDate.time + expiresInSeconds * 1000)
                                auth.expirationDate = expirationDate
                            }
                        }
                    }
                }
            }
        }
        return auth
    }
}
